File: 백준/Gold/17298. 오큰수/오큰수.cc
#include <iostream>
#include <map>
#include <stack>

#define SIZE 1000001
using namespace std;
int N;
int x;
int arr[SIZE];
stack<int> temp;
int result[SIZE];

int main(void) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> N;

    for (int i = 0; i < N; i++) {
        cin >> arr[i];
        result[i] = -1;
    }
    for (int i = 0; i < N; i++) {
        while (!temp.empty() && arr[temp.top()] < arr[i]) {
            result[temp.top()] = arr[i];
            temp.pop();
        }
        temp.push(i);
    }
    for (int i = 0; i < N; i++)
        cout << result[i] << "\n";

    return 0;
}

File: 백준/Gold/17298. 오큰수/오큰수.py
import sys

n = int(sys.stdin.readline().rstrip())
arr = list(map(int, sys.stdin.readline().split()))

answer = [-1] * n
stack = []

stack.append(0)

for i in range(1, n):
    while stack and arr[stack[-1]] < arr[i]:
            answer[stack.pop()] = arr[i]
    stack.append(i)
        
        

print(*answer)